src/Microsoft.SqlTools.ServiceLayer/DisasterRecovery/DisasterRecoveryService.cs

```diff
@@ -310,128 +338,119 @@ private bool IsBackupRestoreOperationSupported(RestoreParams restoreParams, out
         {
-            sqlTask.AddMessage(SR.Task_InProgress, SqlTaskStatus.InProgress, true);
-            IBackupOperation backupOperation = sqlTask.TaskMetadata.Data as IBackupOperation;
-            TaskResult taskResult = null;
-
-            if (backupOperation != null)
-            {
-                AutoResetEvent backupCompletedEvent = new AutoResetEvent(initialState: false);
-                Task<TaskResult> performTask = PerformTaskAsync(backupOperation);
-                Task<TaskResult> cancelTask = CancelTaskAsync(backupOperation, sqlTask.TokenSource.Token, backupCompletedEvent);
-                Task<TaskResult> completedTask = await Task.WhenAny(performTask, cancelTask);
-
-                // Release the cancelTask
-                if (completedTask == performTask)
-                {
-                    backupCompletedEvent.Set();
-                }
-
-                sqlTask.AddMessage(completedTask.Result.TaskStatus == SqlTaskStatus.Failed ? completedTask.Result.ErrorMessage : SR.Task_Completed,
-                                   completedTask.Result.TaskStatus);
-                taskResult = completedTask.Result;
-            }
-            else
-            {
-                taskResult = new TaskResult();
-                taskResult.TaskStatus = SqlTaskStatus.Failed;
-            }
-
-            return taskResult;
+            backupOperation.Execute(TaskExecutionMode.Script);
         }
 
         /// <summary>
         /// Async task to execute backup
         /// </summary>
-        /// <param name="backupOperation"></param>
+        /// <param name="sqlTask"></param>
         /// <returns></returns>
-        private async Task<TaskResult> PerformTaskAsync(IBackupOperation backupOperation)
+        internal async Task<TaskResult> PerformBackupTaskAsync(SqlTask sqlTask)
         {
+            sqlTask.AddMessage(SR.TaskInProgress, SqlTaskStatus.InProgress, true);
+            IBackupOperation backupOperation = sqlTask.TaskMetadata.Data as IBackupOperation;
+            TaskResult result = new TaskResult();
+            string script = "";
+
             // Create a task to perform backup
-            return await Task.Factory.StartNew(() =>
+            await Task.Factory.StartNew(() =>
             {
-                TaskResult result = new TaskResult();
-                try
+                if (backupOperation != null)
                 {
-                    backupOperation.PerformBackup();
-                    result.TaskStatus = SqlTaskStatus.Succeeded;
+                    try
+                    {
+                        backupOperation.Execute(sqlTask.TaskMetadata.TaskExecutionMode);
+                        result.TaskStatus = SqlTaskStatus.Succeeded;
+                        if (!String.IsNullOrEmpty(backupOperation.ScriptContent))
+                        {
+                            // Send generated script to client
+                            sqlTask.AddScript(result.TaskStatus, script);
```

Comment:
This is nice. One minor thing (not required) - it's a good case for just hooking up an event on the BackupOperation that calls sqlTask.AddScript. That way you don't need property checks / more knowledge about internals than is necessary. Instead you just act as a middleman and connect the operation to the task service